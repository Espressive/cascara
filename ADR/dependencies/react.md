# React

Our React version is tied to FE's version, we can bump only after they are doing bumping Ract. For now, all the following ecosystem members will be pinned to their current version:

- @types/react v16.8.0
- @types/react-dom v16.8.0
- react v16.13.1
- react-dom v16.13.1
- react-router v5.2.0
- react-router-dom v5.2.0
- react-scripts v4.0.3
- react-test-renderer v16.13.1
