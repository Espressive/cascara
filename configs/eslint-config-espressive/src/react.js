// "off" or 0 - turn the rule off
// "warn" or 1 - turn the rule on as a warning (doesn't affect exit code)
// "error" or 2 - turn the rule on as an error (exit code will be 1)

module.exports = {
  rules: {
    'react/boolean-prop-naming': 2,
    'react/default-props-match-prop-types': 2,
    'react/destructuring-assignment': 2,
    'react/jsx-fragments': 2,
    'react/jsx-no-duplicate-props': 2,
    'react/jsx-no-useless-fragment': 2,
    'react/jsx-pascal-case': 2,
    'react/jsx-sort-default-props': 2,
    'react/jsx-sort-props': 2,
    'react/no-children-prop': 2,
    'react/no-danger': 2,
    'react/no-find-dom-node': 2,
    'react/no-multi-comp': 2,
    'react/no-string-refs': 2,
    'react/no-this-in-sfc': 2,
    'react/no-typos': 2,
    'react/no-unused-prop-types': 2,
    'react/no-unused-state': 2,
    'react/prefer-stateless-function': 2,
    'react/self-closing-comp': 2,
    'react/sort-prop-types': 2,
    'react/style-prop-object': 2,
  },
};
