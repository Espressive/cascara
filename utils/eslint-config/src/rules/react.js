const { fixtures } = require('../utils/filePaths');

// For brevity, we start with the recomended defaults and extend with our own rules.
// Please make sure we are looking at any rules in the recommended set before adding
// them below. If we are overriding a recommended rule, we should make a comment for
// each rule.
//
// https://github.com/yannickcr/eslint-plugin-react

module.exports = {
  extends: ['plugin:react/recommended'],
  overrides: [
    {
      files: fixtures,
      rules: {
        'react/no-multi-comp': 'off',
        'react/prop-types': 'off',
      },
    },
  ],
  rules: {
    'react/boolean-prop-naming': 'error',
    'react/button-has-type': 'error',
    'react/default-props-match-prop-types': 'error',
    'react/destructuring-assignment': 'error',
    'react/forbid-foreign-prop-types': 'error',
    'react/forbid-prop-types': 'error',
    'react/function-component-definition': [
      'error',
      {
        namedComponents: 'arrow-function',
        unnamedComponents: 'arrow-function',
      },
    ],
    'react/jsx-boolean-value': 'error',
    'react/jsx-fragments': 'error',
    'react/jsx-key': 'error',
    'react/jsx-no-bind': 'error',
    'react/jsx-no-duplicate-props': 'error',
    'react/jsx-no-target-blank': 'error',
    'react/jsx-no-useless-fragment': 'error',
    'react/jsx-pascal-case': 'error',
    'react/jsx-sort-default-props': 'error',
    'react/jsx-sort-props': 'error',
    // 'react/jsx-wrap-multilines': ['error', {
    //   "arrow": "parens-new-line",
    //   "assignment": "parens-new-line",
    //   "condition": "ignore",
    //   "declaration": "parens-new-line",
    //   "logical": "ignore",
    //   "prop": "ignore",
    //   "return": "parens-new-line",
    // }],
    'react/no-access-state-in-setstate': 'error',
    'react/no-children-prop': 'error',
    'react/no-danger': 'error',
    'react/no-find-dom-node': 'error',
    'react/no-multi-comp': 'error',
    'react/no-render-return-value': 'error',
    'react/no-string-refs': 'error',
    'react/no-this-in-sfc': 'error',
    'react/no-typos': 'error',
    'react/no-unused-prop-types': 'error',
    'react/no-unused-state': 'error',
    'react/prefer-stateless-function': 'error',
    'react/require-render-return': 'error',
    'react/self-closing-comp': 'error',
    'react/sort-prop-types': 'error',
    'react/style-prop-object': 'error',
  },
};
