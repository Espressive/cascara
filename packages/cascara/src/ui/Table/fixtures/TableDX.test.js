import { render, screen, within } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import cosmosFixtures, { dataResults } from './TableDX.fixture';

const {
  Loading: loadingFixture,
  Empty: emptyFixture,
  'Data Only': dataOnlyFixture,
  'Data w/ dataDisplay': dataWithDisplayFixture,
} = cosmosFixtures;
const RESULTS_KEYS = Object.keys(dataResults[0]);

describe('Table DX', () => {
  test(`should show a loading state if the 'data' prop is 'undefined' or 'null'`, () => {
    const { baseElement } = render(loadingFixture);
    // This technically captures the entire fixture in a snapshot. We should consider if we want to only select the element within for a snapshot or not.
    expect(baseElement).toMatchSnapshot();
  });

  test(`should show an empty state if data is defined but has a length of 0`, () => {
    const { baseElement } = render(emptyFixture);
    expect(baseElement).toMatchSnapshot();
  });

  describe(`should render all 'data' into columns if no other props are defined:`, () => {
    test(`all table headers`, () => {
      render(dataOnlyFixture);
      // Search for the text in the first header column and get the row, which is the table header
      const tableHeader = within(
        screen.getByText(RESULTS_KEYS[0]).closest('tr')
      );

      // Expect each header in the results keys is rendered in the document
      RESULTS_KEYS.forEach((key) => {
        expect(tableHeader.getByText(key)).toBeInTheDocument();
      });
    });

    test(`all rows render`, () => {
      render(dataOnlyFixture);
      // Test only within the table body, since there are also
      // rows and cells in the table header
      const tableBody = within(
        screen.getByRole('table').querySelector('tbody')
      );

      // get all of the rows in the table body
      const rows = tableBody.getAllByRole('row');

      // Expect that we have rendered the same number of rows as we have data results
      expect(rows).toHaveLength(dataResults.length);
      // Expect that we have the same number of cells in each row as we have keys in our data objects
      rows.forEach((row) =>
        expect(within(row).getAllByRole('cell')).toHaveLength(
          RESULTS_KEYS.length
        )
      );
    });
  });

  test(`should render 'data' from 'dataDisplay' configured settings with no other props defined`, () => {
    render(dataWithDisplayFixture);

    // Check to make sure we only have the number of columns as defined in dataDisplay.
    // This is only checking column headers, not every row.
    expect(screen.getAllByRole('columnheader')).toHaveLength(
      dataWithDisplayFixture.props.dataDisplay.length
    );
  });
});
