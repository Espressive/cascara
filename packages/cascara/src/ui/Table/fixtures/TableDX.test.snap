// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table DX should show a loading state if the 'data' prop is 'undefined' or 'null' 1`] = `
<body>
  <div>
    <h3>
      Loading
    </h3>
    <p>
      The table component will not blow up if there are no values defined. It will also explicitly show a loading state if the data attribute is undefined (or null).
    </p>
    <div
      class="ui segment"
      style="height: 10em;"
    >
      <div
        class="ui active inverted dimmer"
      >
        <div
          class="ui text loader"
        >
          Loading
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Table DX should show an empty state if data is defined but has a length of 0 1`] = `
<body>
  <div>
    <h3>
      Empty
    </h3>
    <p>
      If a table gets an empty array, that should be treated different from undefined or null. An empty array should mean that there are no records and we should indicate this to users in the table itself.
    </p>
    <div
      class="ui placeholder segment"
      style="align-items: center; display: flex; height: 10em; justify-content: center;"
    >
      <em>
        No results...
      </em>
    </div>
  </div>
</body>
`;
